### ISSUE TITLE
Investigate compatibility with Civtas City States Mod

### ISSUE THREAD

**Issue by the-m4a** [civfanatics/CQUI_Community-Edition#101]
We've had enough requests that it's at least worth looking into.  That mod looks interesting, so I'm curious.

**Comment by Infixo** [comment]
Civitas’ mod adds a) new types of CS like maritime, agricultural, etc. b) a new level of suzerain bonuses when you have 10 envoys. To support this they need an updated UI window which they created and it comes with the mod.
There is nothing for CQUi to do here - civitas window must load after cqui and overwrite cqui’s window. It is a matter of load order which is too low atm in their mod and hence the workaround - to delete cqui’s files.

**Comment by alimulhuda** [comment]
I looked into it for a bit in issue #16 . It's pretty much what Infixo says, they should be the ones to increase load order and things should work then.

**Comment by the-m4a** [comment]
The only things CQUI was doing with the CityStates files was to add the "2nd place" Envoy , so we have a full file replacement for both lua and xml to accomplish that.  I want to get back to working on the "extend, not replace" thing so this was a good candidate to look at. 

I came up with a decent compromise that allows CIVITAS to work in parallel - we lose the permanent 2nd place showing on the CityState info row, but gain a tooltip which shows everyone with any Envoys sent to that CS.   

CIVITAS replaces that stuff for their 10 Envoy thing, and have a load-order of 50, which means we need to ensure we load later than them in order to extend the AddCityStateRow function, so the tooltip can be added to the button showing the number of envoys  (it can go anywhere, but I figured the number already sent was a decent place for it).

Showing an existing game (screenshot doesn't get my mouse, but I am pointing to the 15 on Mitla):
![image](https://user-images.githubusercontent.com/8787640/88446899-b7e89f00-cde2-11ea-8e24-ffa7d41bc419.png)

With CIVITAS mod running:
![image](https://user-images.githubusercontent.com/8787640/88446857-6c35f580-cde2-11ea-80aa-f6a18ed10cde.png)

I'll send in a PR once we're sure the stuff with the July patch is good.

**Comment by alimulhuda** [comment]
I've merged the fix for the July patch, so you should be good to go. When I was looking into this issue I think I saw another conflict with the trade overview screen, that's the one to the right of the city state screen. Did you see any issues with that screen?

**Comment by the-m4a** [comment]
Not that I saw, but didn't look in-depth.  Will check it out once I get the files all merged

**Comment by Infixo** [comment]
If the only added value provided by CQUI is a single tooltip then I vote to completely remove this file from the mod. It is not worth it. That way other mods can have a go at CityStates screen without conflict.
I use a screen from ConciseUI which:
- shows CS quest at a glance - super helpful
- shows a compacted, neat summary of all CSs
- has removed bonus section and has more space for CS - screen looks much better
I also added
- information about placed Spy, and it it is traveling
- information about resources that can be acquired by making a CS suzerain
 All this can be added to CQUI.

**Comment by the-m4a** [comment]
It's already done - see the PR.  The changes were not really much.  ConciseUI is another mod?  I would be curious to see what that looks like, it sounds interesting.

**Comment by Infixo** [comment]
I’ve seen the PR. “It’s already done” is not an argument when discussing what should be done in the first place.
And the discussion here how to make Civitas mod compatible with CQUI. This is possible without changing anything in CQUI.
So why change anything? Especially when this change makes CQUI a bit worse. Our primary goal is to make CQUI better. We can work on compatibility but not at an expense of the CQUI itself.

**Comment by the-m4a** [comment]
Was replying there, but this is the better place for the discussion:

CQUI is what should change, in part because as-is, it violates the argument that you made in the other thread that as a UI-overhaul mod, it should load early.

Right now (version on the Workshop), the load order for an EXP2 game is as follows:
- CQUI citystates.lua: 0 (assets/ui/partialscreens/citystates.lua)
- Civitas citystates.lua: 51 (via their CityStates_Expansion1.lua file)
- Civitas citystates.xml: 51
- CQUI citystates.xml: 200 (assets/expansion2/replacements/citystates.xml)

There doesn't appear to be a significant reason that the CQUI Expansion2 stuff is loading at 200.  The actual EXP2 content from Firaxis has no LoadOrder values mentioned in the EXP2 modinfo file, so this 200 in CQUI appears to be more the problem and where Civitas is loading.

I changed the LO values in the CQUI.modinfo to 1, 2, and 3 respectively for the Expansion1, Expansion1AndBeyond, and Expansion2 ImportFiles declarations (previously, they were 100, 150, and 200).  I loaded the game with Civitas up and it functions correctly.  Obviously, because we have so many other files that were using these LO values, there is more testing to be done, but this is what we should be looking at.   At least for this pairing of files, it doesn't make sense to have 1 use load-order 0 and the other to use load-order 200.

As an aside, something I noticed while looking at this stuff:
I don't see any errors in the FireTuner when CQUI/Civitas are in the broken state.  When using our current load order, this is the result with Civitas:
![image](https://user-images.githubusercontent.com/8787640/88452070-87b7f500-ce10-11ea-87f7-d43da826b84c.png)

Commenting out the references to the "Bonusxxx10" objects (total of 5 lines) in AddCityStateRow of the Civitas citystates.lua allows the full panel to populate again.  These objects are not defined in the CQUI citystates.xml, obviously.
Note there is no CQUI 2nd Suzerain, as there's nothing to populate that information:
![image](https://user-images.githubusercontent.com/8787640/88452153-2e9c9100-ce11-11ea-8873-bd210c1d9552.png)

**Comment by the-m4a** [comment]
Okay, trying again.  Maintains functionality and adds the tooltip.
![image](https://user-images.githubusercontent.com/8787640/88453794-319e7e00-ce1f-11ea-83e4-55d961e1671e.png)

![image](https://user-images.githubusercontent.com/8787640/88454579-893fe800-ce25-11ea-8000-144fb372d034.png)

**Comment by Infixo** [comment]
I repeat the question - why do you want to change ANYTHING in CQUI? There is no need to do so.
Civitas can make their mod compatible with CQUI. If they DON’T want, I see no reason why should we care at all?

 “Loading first” does not mean 1, 2, 3 - it means that all other mods that want to compatible with CQUI need to load after CQUI. And LO values used in CQUI are pretty good, aligned with Firaxis values and also flexible enough so other mods can put things in between if they want.

**Comment by alimulhuda** [comment]
I agree with Infixo here. The load order we have in CQUI is very nicely spaced out and allows other mods to load their code where they want, I am opposed to reducing the load order specially to something as low as single digits. Given that there are already mods that work around CQUIs current load order I would be hesitant to change any load order stuff. I also very much prefer having the "second place at a glance" feature that was temporarily removed.

If you look at other mods that aim to be compatible with CQUI, like astog's and Infixo's stuff I believe, then you'll see they generally adjust their load order to work nicely. Personally my preferred action here would be one of the following (highest preference first)

- Do nothing. Anytime someone asks about CIVITAS tell them to contact CIVITAS and ask to increase their load order. Also there are workarounds to make it work with CQUI.
- A pure modinfo update where we just don't load the conflicting files if we detect CIVITAS (same effect as deleting the CQUI files, but only takes effect if CIVITAS is detected), but I don't know if modinfo has a way to do that
- Make changes without touching load order if possible

**Comment by the-m4a** [comment]
Why change CQUI?  Because I am not certain CQUI is doing the right thing, and I'd prefer we are certain that CQUIs behavior is correct before asking/expecting others to modify their mods.  

I agree that our overall goal is to maintain and improve this mod, and I think there's plenty of evidence to suggest that design decisions made *before* the three of us started to work on this in earnest may not have been the best, so therefore it is reasonable for any one of us to question those decisions.

I also want to express that I very much appreciate this discussion and the input from you both, because I think the end result is a better mod, or at least a better understanding of how it should all work (even if only for me).

The question:
When starting a game with RF and/or GS loaded, should CQUI be loading CityStates.xml at 200?

My observations are:
- CQUI and Civitas work together in the vanilla game (no RF, no GS), are broken when expansions are loaded
- CQUI _always_ loads citystates.lua at loadorder 0, but citystates.xml  varies: 0 for vanilla, 100 for RF, 200 for GS.
  - Civitas loads citystates.lua at 50, the xml at 50 for vanilla, 51 for expansions 1 and 2.
- The Firaxis Expansion1.modinfo and Expansion2.modinfo specify no LoadOrder values, so the 200 used by CQUI may be arbitrary and unnecessarily high
- The documentation on LoadOrder - questionable as it may be, because Firaxis doesn't appear to have updated it for the expansions - leaves room for interpretation and that's the root of this confusion.

From the doc in the Civ6 SDK:
|Value|Description|
|-|-|
|-100|Schema changes, Remove Data, Populate core data| 
|-75|Unofficial Schema changes|
|-50|Premature Updates: Updates that need to be made before content is typically added.|
|-25|Unofficial changes to premature updates|
|0|Standard Content Updates: Most actions applied during this phase.|
|25|Unofficial changes to standard updates| 
|50|Post Updates: Updates that require most content to be provided.|
|75|Unofficial changes to post-update content| 
|100|Scenario Updates: Often applied at the very end of everything else. Called out as its own phase.|
|125|Unofficial changes to Scenario updates.|

I posted a comment on the CIVITAS mod page asking his reasoning for using 50/51.  I'm curious if the chart above was the guide, or if 50/51 was as arbitrary as our 0/100/200 appears to be.

**Comment by the-m4a** [comment]
also updated the PR to decouple the "extend, not replace" from "make work with civitas", so at least that part can move forward.  I like the tooltip. :P

**Comment by the-m4a** [comment]
SeelingCat sent me a msg on steam, says he'll adjust.  I am still not sure CQUI is doing the right thing, but will work with him and get it done.

**Comment by the-m4a** [comment]
CIVITAS has been updated.  I'll likewise push an update.  We'll call this good.
